File: python_code/bin/pullUp_resistor.py
import RPi.GPIO as GPIO
import time

# Setup
GPIO.setwarnings(False)
GPIO.setmode(GPIO.BOARD)

# Define pins
BUTTON_PIN = 40  # Your button is on pin 40

# Setup pin with internal pull-up resistor
# This means the input will read HIGH (1) when the button is NOT pressed
# and LOW (0) when the button IS pressed (connecting to ground)
GPIO.setup(BUTTON_PIN, GPIO.IN, pull_up_down=GPIO.PUD_UP)

def button_callback(channel):
    if GPIO.input(BUTTON_PIN) == GPIO.LOW:
        print("Button pressed! (LOW)")
    else:
        print("Button released! (HIGH)")

# Register interrupt for both rising and falling edge
GPIO.add_event_detect(BUTTON_PIN, GPIO.BOTH, callback=button_callback, bouncetime=200)

try:
    print("Push button monitoring started. Press Ctrl+C to exit.")
    print("Current state: {} ({})".format(
        "HIGH" if GPIO.input(BUTTON_PIN) == GPIO.HIGH else "LOW",
        "Button NOT pressed" if GPIO.input(BUTTON_PIN) == GPIO.HIGH else "Button pressed"
    ))
    
    while True:
        time.sleep(0.1)  # Just keep the program running

except KeyboardInterrupt:
    print("\nProgram stopped by user")
finally:
    GPIO.cleanup()  # Clean up GPIO on exit

File: rust_code/pi_afr/src/bin/pullUp_resistor.rs
use rppal::gpio::{Gpio, InputPin, Level, Trigger};
use std::error::Error;
use std::sync::{Arc, atomic::{AtomicBool, Ordering}};
use std::thread;
use std::time::Duration;

fn main() -> Result<(), Box<dyn Error>> {
    // BCM pin 21 = Physical pin 40
    let button_pin_bcm = 21; 
    
    // Initialize GPIO
    let gpio = Gpio::new()?;
    
    // Set up the button pin with internal pull-up resistor
    // This means the input will read High when the button is NOT pressed
    // and Low when the button IS pressed (connecting to ground)
    let mut button = gpio.get(button_pin_bcm)?.into_input_pullup();
    
    // Set up interrupt handling
    let running = Arc::new(AtomicBool::new(true));
    let r = running.clone();
    
    // Handle Ctrl+C
    ctrlc::set_handler(move || {
        r.store(false, Ordering::SeqCst);
    })?;
    
    // Display initial state
    println!("Push button monitoring started. Press Ctrl+C to exit.");
    println!("Current state: {} ({})",
             if button.is_high() { "HIGH" } else { "LOW" },
             if button.is_high() { "Button NOT pressed" } else { "Button pressed" });
    
    // Set up interrupt for both rising and falling edge
    button.set_interrupt(Trigger::Both)?;
    
    while running.load(Ordering::SeqCst) {
        match button.poll_interrupt(true, Some(Duration::from_millis(100)))? {
            Some(level) => {
                match level {
                    Level::Low => println!("Button pressed! (LOW)"),
                    Level::High => println!("Button released! (HIGH)"),
                }
            },
            None => {} // Timeout, do nothing
        }
    }
    
    println!("Program completed");
    
    Ok(())
}
